File: network/src/main/java/com/github/vivchar/network/ForksManager.java
package com.github.vivchar.network;

import androidx.annotation.NonNull;

import com.github.vivchar.network.models.GithubFork;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.subjects.ReplaySubject;

/**
 * Created by Vivchar Vitaly on 12.10.17.
 */

public class ForksManager {

	@NonNull
	private final GithubClient mClient;
	@NonNull
	private final ReplaySubject<List<GithubFork>> mGithubForks = ReplaySubject.createWithSize(1);

	public ForksManager(@NonNull final GithubClient client) {
		mClient = client;
	}

	public void onApplicationStarted() {
		mClient.sendForksRequest();
	}

	public void onForksReceived(@NonNull final List<GithubFork> forks) {
		mGithubForks.onNext(forks);
	}

	public void onForksFailed(@NonNull final String message) {
		mGithubForks.onNext(new ArrayList<>());
//		mGithubForks.onError(new Exception(message));
	}

	@NonNull
	public Observable<List<GithubFork>> getGithubForks() {
		return mGithubForks.hide();
	}
}

File: network/src/main/java/com/github/vivchar/network/ForksManager.kt
package com.github.vivchar.network

import com.github.vivchar.network.models.GithubFork
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.subjects.ReplaySubject

/**
 * Created by Vivchar Vitaly on 12.10.17.
 */
class ForksManager internal constructor(private val client: GithubClient) {

	private val githubForksSubject = ReplaySubject.createWithSize<List<GithubFork>>(1)

	fun onApplicationStarted() {
		client.sendForksRequest()
	}

	fun onForksReceived(forks: List<GithubFork>) {
		githubForksSubject.onNext(forks)
	}

	fun onForksFailed(message: String) {
		githubForksSubject.onNext(ArrayList())
//		mGithubForks.onError(new Exception(message));
	}

	val githubForks: Observable<List<GithubFork>> get() = githubForksSubject.hide()
}